File: app/src/main/java/com/mmjang/ankihelper/ui/LauncherActivity.java
package com.mmjang.ankihelper.ui;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.CompoundButton;
import com.google.android.material.materialswitch.MaterialSwitch;
import android.widget.TextView;
import android.widget.Toast;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Spinner;

import com.mmjang.ankihelper.R;
import com.mmjang.ankihelper.anki.AnkiDroidHelper;
import com.mmjang.ankihelper.data.database.DatabaseManager;
import com.mmjang.ankihelper.data.database.MigrationUtil;
import com.mmjang.ankihelper.data.plan.DefaultPlan;
import com.mmjang.ankihelper.data.plan.OutputPlanPOJO;
import com.mmjang.ankihelper.domain.CBWatcherService;
import com.mmjang.ankihelper.MyApplication;
import com.mmjang.ankihelper.data.Settings;
import com.mmjang.ankihelper.data.AppTheme;
import com.mmjang.ankihelper.data.ThemeManager;


import com.mmjang.ankihelper.ui.plan.PlansManagerActivity;
import com.mmjang.ankihelper.ui.stat.StatActivity;


import java.util.List;
import java.util.ArrayList;

public class LauncherActivity extends AppCompatActivity {

    AnkiDroidHelper mAnkiDroid;
    Settings settings;
    DatabaseManager databaseManager;

    //views
    MaterialSwitch switchMoniteClipboard;
    MaterialSwitch switchCancelAfterAdd;
    MaterialSwitch switchLeftHandMode;
    Spinner themeSpinner;
    TextView textViewOpenPlanManager;
    TextView textViewOpenAIConfig;
    TextView textViewAddDefaultPlan;
    TextView textViewAcknowledge;
    TextView textViewOpenStatistics;

    private static final int REQUEST_CODE_ANKI = 0;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        settings = Settings.getInstance(LauncherActivity.this);
        ThemeManager.applyTheme(this);
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_launcher); // Set the layout first
        setVersion();

        // Initialize AnkiDroidHelper in onCreate
        mAnkiDroid = MyApplication.getAnkiDroid(this);

        // Call checkAndRequestPermissions after AnkiDroidHelper is initialized
        checkAndRequestPermissions();

        // Calculate the database path

        // Initialize DatabaseManager after setting the layout
        databaseManager = databaseManager.getInstance();

        switchMoniteClipboard = findViewById(R.id.switch_monite_clipboard);
        switchCancelAfterAdd = findViewById(R.id.switch_cancel_after_add);
        switchLeftHandMode = findViewById(R.id.left_hand_mode);
        themeSpinner = findViewById(R.id.theme_spinner);
        textViewOpenPlanManager = (TextView) findViewById(R.id.btn_open_plan_manager);
        textViewOpenAIConfig = (TextView) findViewById(R.id.btn_open_ai_config);
        textViewAddDefaultPlan = (TextView) findViewById(R.id.btn_add_default_plan);
        textViewAcknowledge = (TextView) findViewById(R.id.textview_acknowledge);
        textViewOpenStatistics = (TextView) findViewById(R.id.btn_open_statistics);

        switchMoniteClipboard.setChecked(
                settings.getMoniteClipboardQ()
        );

        switchCancelAfterAdd.setChecked(
                settings.getAutoCancelPopupQ()
        );

        switchLeftHandMode.setChecked(
                settings.getLeftHandModeQ()
        );

        // Setup theme spinner
        setupThemeSpinner();

        switchMoniteClipboard.setOnCheckedChangeListener(
                new CompoundButton.OnCheckedChangeListener() {
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                        settings.setMoniteClipboardQ(isChecked);
                        if (isChecked) {
                            startCBService();
                        } else {
                            stopCBService();
                        }
                    }
                }
        );

        switchLeftHandMode.setOnCheckedChangeListener(
                new CompoundButton.OnCheckedChangeListener() {
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                        settings.setLeftHandModeQ(isChecked);
                    }
                }
        );

        switchCancelAfterAdd.setOnCheckedChangeListener(
                new CompoundButton.OnCheckedChangeListener() {
                    @Override
                    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                        settings.setAutoCancelPopupQ(isChecked);
                    }
                }
        );

        
        textViewOpenPlanManager.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (!mAnkiDroid.isAnkiDroidRunning()) {
                    Toast.makeText(LauncherActivity.this, R.string.api_not_available_message, Toast.LENGTH_LONG).show();
                    return;
                }

                if (mAnkiDroid.shouldRequestPermission()) {
                    mAnkiDroid.requestPermission(LauncherActivity.this, 0);
                    return;
                }

                Intent intent = new Intent(LauncherActivity.this, PlansManagerActivity.class);
                startActivity(intent);
            }
        });

        textViewOpenAIConfig.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LauncherActivity.this, com.mmjang.ankihelper.ui.ai.AIConfigActivity.class);
                startActivity(intent);
            }
        });

        textViewAddDefaultPlan.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (!AnkiDroidHelper.isApiAvailable(MyApplication.getContext())) {
                            Toast.makeText(LauncherActivity.this, R.string.api_not_available_message, Toast.LENGTH_LONG).show();
                            return;
                        }

                        if (mAnkiDroid.shouldRequestPermission()) {
                            mAnkiDroid.requestPermission(LauncherActivity.this, 0);
                            return;
                        } else {

                        }
                        askIfAddDefaultPlan();
                    }
                }
        );
        
        textViewOpenStatistics.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(LauncherActivity.this, StatActivity.class);
                startActivity(intent);
            }
        });
        //debug new feature
//        Thread thread = new Thread(new Runnable() {
//            @Override
//            public void run() {
//                try {
//                    YoudaoOnline.getDefinition("dedicate");
//                }
//                catch (IOException e){
//
//                }
//            }
//        });
//        thread.start();
    }
    private void checkAndRequestPermissions() {
      if (mAnkiDroid == null) {
          mAnkiDroid = new AnkiDroidHelper(this);
      }
      if (mAnkiDroid.shouldRequestPermission()) {
          mAnkiDroid.requestPermission(this, REQUEST_CODE_ANKI);
      }
        // Only check notification permission (for internal storage)
        if (Build.VERSION.SDK_INT >= 23 &&
            ContextCompat.checkSelfPermission(this, android.Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{android.Manifest.permission.POST_NOTIFICATIONS}, REQUEST_CODE_ANKI);
            return;
        }
    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }


    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        // Call super.onRequestPermissionsResult() first
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (grantResults.length == 0) {
            return;
        }

        if (requestCode == REQUEST_CODE_ANKI ) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ensureExternalDbDirectoryAndMigrate();
            } else {
                //Toast.makeText(this, R.string.permission_denied, Toast.LENGTH_LONG).show();
                new AlertDialog.Builder(LauncherActivity.this)
                        .setMessage(R.string.permission_denied)
                        .setIcon(android.R.drawable.ic_dialog_alert)
                        .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                openSettingsPage();
                            }
                        }).show();
            }
        }
        // The REQUEST_CODE_STORAGE is no longer used for internal storage
        // ...
    }

    private void ensureExternalDbDirectoryAndMigrate() {
        // Check for storage permission
        //if (Build.VERSION.SDK_INT >= 23 && ContextCompat.checkSelfPermission(this, android.Manifest.permission.WRITE_EXTERNAL_STORAGE) != PackageManager.PERMISSION_GRANTED) {
        //    // Request permission
        //    ActivityCompat.requestPermissions(this, new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, REQUEST_CODE_STORAGE);
        //    return; // Exit the function to wait for permission result
        //}
        // ... the rest of the function is no longer needed, as we are using internal storage now
    }

    private void startCBService() {
        Intent intent = new Intent(this, CBWatcherService.class);
        startService(intent);
    }

    private void stopCBService() {
        Intent intent = new Intent(this, CBWatcherService.class);
        stopService(intent);
    }

    void askIfAddDefaultPlan() {
        List<OutputPlanPOJO> plans;
        plans = databaseManager.getAllPlan(); // Access internal database by default

        for (OutputPlanPOJO plan : plans) {
            if (plan.getPlanName().equals(DefaultPlan.DEFAULT_PLAN_NAME)) {
                new AlertDialog.Builder(LauncherActivity.this)
                        .setMessage(R.string.duplicate_plan_name_complain)
                        .setIcon(android.R.drawable.ic_dialog_alert)
                        .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                            public void onClick(DialogInterface dialog, int whichButton) {
                                return;
                            }
                        }).show();
                return;
            }
        }
        if (plans.size() == 0) {
            new AlertDialog.Builder(LauncherActivity.this)
                    .setTitle(R.string.confirm_add_default_plan)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int whichButton) {
                            DefaultPlan plan = new DefaultPlan(LauncherActivity.this);
                            plan.addDefaultPlan();
                            Toast.makeText(LauncherActivity.this, R.string.default_plan_added, Toast.LENGTH_SHORT).show();
                        }
                    })
                    .setNegativeButton(android.R.string.no, null).show();
        } else {
            new AlertDialog.Builder(LauncherActivity.this)
                    .setMessage(R.string.confirm_add_default_plan_when_exists_already)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                        public void onClick(DialogInterface dialog, int whichButton) {
                            try {
                                DefaultPlan plan = new DefaultPlan(LauncherActivity.this);
                                plan.addDefaultPlan();
                                Toast.makeText(LauncherActivity.this, R.string.default_plan_added, Toast.LENGTH_SHORT).show();
                            }catch (Exception e){
                                Toast.makeText(LauncherActivity.this, e.getLocalizedMessage(), Toast.LENGTH_SHORT).show();
                            }
                        }
                    })
                    .setNegativeButton(android.R.string.no, null).show();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
    }

    private void openSettingsPage() {
        Intent intent = new Intent(this, com.mmjang.ankihelper.ui.settings.SettingsActivity.class);
        startActivity(intent);
    }

    /**
     * Setup theme selection spinner
     */
    private void setupThemeSpinner() {
        // Create adapter with theme display names
        String[] themeNames = new String[AppTheme.values().length];
        for (int i = 0; i < AppTheme.values().length; i++) {
            themeNames[i] = AppTheme.values()[i].getDisplayName();
        }

        ArrayAdapter<String> themeAdapter = new ArrayAdapter<>(this,
                R.layout.custom_spinner_item, themeNames);
        themeAdapter.setDropDownViewResource(R.layout.custom_spinner_dropdown_item);
        themeSpinner.setAdapter(themeAdapter);

        // Set current selection
        AppTheme currentTheme = settings.getSelectedTheme();
        themeSpinner.setSelection(currentTheme.ordinal());

        // Handle theme selection changes
        themeSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
                AppTheme selectedTheme = AppTheme.values()[position];
                AppTheme currentTheme = settings.getSelectedTheme();

                if (selectedTheme != currentTheme) {
                    settings.setSelectedTheme(selectedTheme);
                    
                    // Show confirmation dialog for theme change
                    showThemeChangeDialog(selectedTheme);
                }
            }

            @Override
            public void onNothingSelected(AdapterView<?> parent) {
                // Do nothing
            }
        });
    }

    /**
     * Show theme change confirmation dialog
     */
    private void showThemeChangeDialog(AppTheme newTheme) {
        new AlertDialog.Builder(this)
                .setTitle(R.string.theme_change_title)
                .setMessage(getString(R.string.theme_change_message, newTheme.getDisplayName()))
                .setPositiveButton(R.string.apply, (dialog, which) -> {
                    // Apply theme immediately
                    recreate();
                })
                .setNegativeButton(R.string.cancel, (dialog, which) -> {
                    // Revert spinner selection
                    AppTheme currentTheme = settings.getSelectedTheme();
                    themeSpinner.setSelection(currentTheme.ordinal());
                })
                .show();
    }

    public void setVersion() {
        try {
            String versionName = getPackageManager().getPackageInfo(getPackageName(), 0).versionName;
            TextView versionTextView = (TextView) findViewById(R.id.textview_version);
            versionTextView.setText(
                    "Ver: " + versionName
            );
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }

}

File: app/src/main/java/com/lmyby/ankihelper/ui/LauncherActivity.kt
package com.lmyby.ankihelper.ui

import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import com.google.android.material.materialswitch.MaterialSwitch
import com.lmyby.ankihelper.MyApplication
import com.lmyby.ankihelper.R
import com.lmyby.ankihelper.anki.AnkiDroidHelper
import com.lmyby.ankihelper.data.AppLanguage
import com.lmyby.ankihelper.data.AppTheme
import com.lmyby.ankihelper.data.LanguageManager
import com.lmyby.ankihelper.data.Settings
import com.lmyby.ankihelper.data.ThemeManager
import com.lmyby.ankihelper.domain.CBWatcherService
import com.lmyby.ankihelper.ui.ai.AIConfigActivity
import com.lmyby.ankihelper.ui.plan.PlansManagerActivity
import com.lmyby.ankihelper.ui.settings.SettingsActivity
import com.lmyby.ankihelper.ui.stat.StatActivity
import com.lmyby.ankihelper.ui.storage.StorageMigrationActivity
import com.lmyby.ankihelper.util.StorageManager

class LauncherActivity : AppCompatActivity() {

    private lateinit var mAnkiDroid: AnkiDroidHelper
    private lateinit var settings: Settings

    // Views
    private lateinit var switchMoniteClipboard: MaterialSwitch
    private lateinit var switchCancelAfterAdd: MaterialSwitch
    private lateinit var switchLeftHandMode: MaterialSwitch
    private lateinit var themeSpinner: Spinner
    private lateinit var languageSpinner: Spinner
    private lateinit var textViewOpenPlanManager: TextView
    private lateinit var textViewOpenAIConfig: TextView
    private lateinit var textViewAcknowledge: TextView
    private lateinit var textViewOpenStatistics: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        settings = Settings.getInstance(this@LauncherActivity)
        ThemeManager.applyTheme(this)
        // Language is now handled automatically by AndroidX AppCompatDelegate
        // Sync Settings with the current app language on startup
        LanguageManager.syncWithSettings(this)
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_launcher) // Set the layout first
        setVersion()

        // Check for storage migration requirement
        checkStorageMigration()

        // Initialize AnkiDroidHelper in onCreate
        mAnkiDroid = MyApplication.getAnkiDroid(this)

        // Call checkAndRequestPermissions after AnkiDroidHelper is initialized
        checkAndRequestPermissions()

        // Initialize views
        switchMoniteClipboard = findViewById(R.id.switch_monite_clipboard)
        switchCancelAfterAdd = findViewById(R.id.switch_cancel_after_add)
        switchLeftHandMode = findViewById(R.id.left_hand_mode)
        themeSpinner = findViewById(R.id.theme_spinner)
        languageSpinner = findViewById(R.id.language_spinner)
        textViewOpenPlanManager = findViewById(R.id.btn_open_plan_manager)
        textViewOpenAIConfig = findViewById(R.id.btn_open_ai_config)
        textViewAcknowledge = findViewById(R.id.textview_acknowledge)
        textViewOpenStatistics = findViewById(R.id.btn_open_statistics)

        switchMoniteClipboard.isChecked = settings.getMoniteClipboardQ()
        switchCancelAfterAdd.isChecked = settings.getAutoCancelPopupQ()
        switchLeftHandMode.isChecked = settings.getLeftHandModeQ()

        // Setup theme spinner
        setupThemeSpinner()

        // Setup language spinner
        setupLanguageSpinner()

        switchMoniteClipboard.setOnCheckedChangeListener { _, isChecked ->
            settings.setMoniteClipboardQ(isChecked)
            if (isChecked) {
                startCBService()
            } else {
                stopCBService()
            }
        }

        switchLeftHandMode.setOnCheckedChangeListener { _, isChecked ->
            settings.setLeftHandModeQ(isChecked)
        }

        switchCancelAfterAdd.setOnCheckedChangeListener { _, isChecked ->
            settings.setAutoCancelPopupQ(isChecked)
        }

        textViewOpenPlanManager.setOnClickListener {
            if (!mAnkiDroid.isAnkiDroidRunning) {
                Toast.makeText(
                    this@LauncherActivity,
                    R.string.api_not_available_message,
                    Toast.LENGTH_LONG
                ).show()
                return@setOnClickListener
            }

            if (mAnkiDroid.shouldRequestPermission()) {
                mAnkiDroid.requestPermission(this@LauncherActivity, 0)
                return@setOnClickListener
            }

            val intent = Intent(this@LauncherActivity, PlansManagerActivity::class.java)
            startActivity(intent)
        }

        textViewOpenAIConfig.setOnClickListener {
            val intent = Intent(this@LauncherActivity, AIConfigActivity::class.java)
            startActivity(intent)
        }

        textViewOpenStatistics.setOnClickListener {
            val intent = Intent(this@LauncherActivity, StatActivity::class.java)
            startActivity(intent)
        }
    }

    private fun checkStorageMigration() {
        val storageManager = StorageManager(this, getSharedPreferences("ankihelper_prefs", MODE_PRIVATE))

        // Check if migration is needed
        if (!storageManager.isMigrationCompleted()) {
            val intent = Intent(this, StorageMigrationActivity::class.java)
            startActivity(intent)
            finish() // Close launcher activity
        }
    }

    private fun checkAndRequestPermissions() {
        if (mAnkiDroid.shouldRequestPermission()) {
            mAnkiDroid.requestPermission(this, REQUEST_CODE_ANKI)
        }
        // Only check notification permission (for internal storage)
        if (Build.VERSION.SDK_INT >= 23 &&
            ContextCompat.checkSelfPermission(
                this,
                android.Manifest.permission.POST_NOTIFICATIONS
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                this,
                arrayOf(android.Manifest.permission.POST_NOTIFICATIONS),
                REQUEST_CODE_ANKI
            )
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater: MenuInflater = menuInflater
        inflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return super.onOptionsItemSelected(item)
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        // Call super.onRequestPermissionsResult() first
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if (grantResults.isEmpty()) {
            return
        }

        if (requestCode == REQUEST_CODE_ANKI) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                ensureExternalDbDirectoryAndMigrate()
            } else {
                AlertDialog.Builder(this@LauncherActivity)
                    .setMessage(R.string.permission_denied)
                    .setIcon(android.R.drawable.ic_dialog_alert)
                    .setPositiveButton(android.R.string.yes) { _: DialogInterface, _: Int ->
                        openSettingsPage()
                    }
                    .show()
            }
        }
        // The REQUEST_CODE_STORAGE is no longer used for internal storage
    }

    private fun ensureExternalDbDirectoryAndMigrate() {
        // Check for storage permission
        // This function is no longer needed, as we are using internal storage now
    }

    private fun startCBService() {
        val intent = Intent(this, CBWatcherService::class.java)
        startService(intent)
    }

    private fun stopCBService() {
        val intent = Intent(this, CBWatcherService::class.java)
        stopService(intent)
    }

    private fun showNoPlansGuidance() {
        AlertDialog.Builder(this)
            .setTitle(R.string.no_plans_found_title)
            .setMessage(R.string.no_plans_found_message)
            .setPositiveButton(R.string.go_to_plan_manager) { _, _ ->
                val intent = Intent(this, PlansManagerActivity::class.java)
                startActivity(intent)
            }
            .setNegativeButton(R.string.cancel, null)
            .show()
    }

    override fun onStop() {
        super.onStop()
    }

    private fun openSettingsPage() {
        val intent = Intent(this, SettingsActivity::class.java)
        startActivity(intent)
    }

    /**
     * Setup theme selection spinner
     */
    private fun setupThemeSpinner() {
        // Create adapter with theme display names
        val themeNames = Array(AppTheme.values().size) { i ->
            AppTheme.values()[i].getDisplayName(this)
        }

        val themeAdapter = ArrayAdapter(
            this,
            R.layout.centered_spinner_item,
            themeNames
        )
        themeAdapter.setDropDownViewResource(R.layout.centered_spinner_dropdown_item)
        themeSpinner.adapter = themeAdapter

        // Set current selection
        val currentTheme = settings.getSelectedTheme()
        themeSpinner.setSelection(currentTheme.ordinal)

        // Handle theme selection changes
        themeSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                val selectedTheme = AppTheme.values()[position]
                val currentTheme = settings.getSelectedTheme()

                if (selectedTheme != currentTheme) {
                    settings.setSelectedTheme(selectedTheme)
                    applyThemeWithoutRestart()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
                // Do nothing
            }
        }
    }

    /**
     * Apply theme changes without requiring app restart
     * Uses modern Android practices for dynamic theme switching
     */
    private fun applyThemeWithoutRestart() {
        // Get the new theme
        val newTheme = settings.getSelectedTheme()

        // Apply theme to the current activity
        setTheme(newTheme.themeResId)

        // Recreate the activity to apply the new theme
        recreate()
    }

    /**
     * Setup language selection spinner
     */
    private fun setupLanguageSpinner() {
        // Create adapter with language display names
        val languageNames = Array(AppLanguage.values().size) { i ->
            AppLanguage.values()[i].getDisplayName(this)
        }

        val languageAdapter = ArrayAdapter(
            this,
            R.layout.centered_spinner_item,
            languageNames
        )
        languageAdapter.setDropDownViewResource(R.layout.centered_spinner_dropdown_item)
        languageSpinner.adapter = languageAdapter

        // Set current selection
        val currentLanguage = settings.getSelectedLanguage()
        languageSpinner.setSelection(currentLanguage.ordinal)

        // Handle language selection changes
        languageSpinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(
                parent: AdapterView<*>?,
                view: View?,
                position: Int,
                id: Long
            ) {
                val selectedLanguage = AppLanguage.values()[position]
                val currentLanguage = settings.getSelectedLanguage()

                if (selectedLanguage != currentLanguage) {
                    settings.setSelectedLanguage(selectedLanguage)
                    applyLanguageWithoutRestart()
                }
            }

            override fun onNothingSelected(parent: AdapterView<*>?) {
                // Do nothing
            }
        }
    }

    /**
     * Apply language changes using modern AndroidX API
     * The API will automatically recreate all activities with the new language
     */
    private fun applyLanguageWithoutRestart() {
        val selectedLanguage = settings.getSelectedLanguage()

        // Use modern AndroidX API - this will:
        // 1. Persist the language preference automatically
        // 2. Recreate all activities with the new language
        // 3. Update the app name and all system UI
        LanguageManager.setAppLanguage(selectedLanguage)

        // No need to manually recreate - AppCompatDelegate handles it
    }

    private fun setVersion() {
        try {
            val versionName = packageManager.getPackageInfo(packageName, 0).versionName
            val versionTextView = findViewById<TextView>(R.id.textview_version)
            versionTextView.text = "Ver: $versionName"
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
    }

    companion object {
        private const val REQUEST_CODE_ANKI = 0
    }
}
